### DISCUSSION TITLE
Option to not show installed wallets when using custom wallet list

### DISCUSSION THREAD

**Discussion by agnideepGhosh** [rainbow-me/rainbowkit#2127] [Q&A] [answered]
In @rainbow-me/rainbowkit v2, when I am using a custom wallet list as described here - https://www.rainbowkit.com/docs/custom-wallet-list, installed wallets also appear in the modal. Is there a way to only show wallets passed in the custom list in the modal?

**Comment by 0xmkh** [comment, marked as answer]
@agnideepGhosh Yes that's possible. You can set `multiInjectedProviderDiscovery` to `false` in your wagmi config.

Example with `createConfig`.

```tsx
import { connectorsForWallets } from '@rainbow-me/rainbowkit';
import {
  rainbowWallet,
  walletConnectWallet,
} from '@rainbow-me/rainbowkit/wallets';
import { createConfig, http } from 'wagmi';
import { mainnet } from 'wagmi/chains';

const connectors = connectorsForWallets(
  [
    {
      groupName: 'Recommended',
      wallets: [rainbowWallet, walletConnectWallet],
    },
  ],
  {
    appName: 'My RainbowKit App',
    projectId: 'YOUR_PROJECT_ID',
  }
);

const config = createConfig({
  connectors,
  multiInjectedProviderDiscovery: false,
  chains: [mainnet],
  transports: {
    [mainnet.id]: http(),
  },
});
```

Example with `getDefaultConfig`.

```tsx
import '@rainbow-me/rainbowkit/styles.css';

import { getDefaultConfig } from '@rainbow-me/rainbowkit';
import {
  rainbowWallet,
  walletConnectWallet,
} from '@rainbow-me/rainbowkit/wallets';
import { mainnet } from 'wagmi/chains';

const config = getDefaultConfig({
  wallets: [
    {
      groupName: 'Recommended',
      wallets: [rainbowWallet, walletConnectWallet],
    },
  ],
  multiInjectedProviderDiscovery: false,
  chains: [mainnet],
  appName: 'My RainbowKit App',
  projectId: 'YOUR_PROJECT_ID',
});
```

**Reply by agnideepGhosh** [reply]
Thanks @magiziz!

**Reply by agnideepGhosh** [reply]
@magiziz Another behavior is that we also use Solana wallets, when we have a common wallet present to connect on Solana and EVM - if its connected for Solana chain and not for EVM, the connection is still picked up on EVM which did not occur before v2 migration - any changes in implementation on that front?

**Reply by 0xmkh** [reply]
This behaviour is coming from Wagmi directly. RainbowKit / Wagmi doesn't support Solana for now, but we're looking to support it in the future.

**Reply by agnideepGhosh** [reply]
I see, I believe its because we are sharing wallet connectors to wagmi directly and possibly because of : https://github.com/wevm/wagmi/blob/main/packages/core/src/actions/disconnect.ts#L55 introduced in v2, it picks up any active connection. Do you think there is there any temporary workaround to prevent this?

**Reply by 0xmkh** [reply]
You could try using other connector libraries that supports Solana. You can display a button saying "Connect to Solana" and have other connector libraries take care of that instead of RainbowKit.

**Reply by agnideepGhosh** [reply]
Yes I'm doing exactly that, but let's say there is a common wallet which is being used in my solana adapter and as well as in my connector list in wagmi config, even though if the solana connector is used to connect to the solana wallet, wagmi seems to pick up the connection and uses the same wallet for EVM (phantom wallet in this case). Is this behavior normal in v2?

**Reply by 0xmkh** [reply]
Yes this is normal. You can make sure to disconnect the phantom wallet with Wagmi if Solana connector is connected or vise versa.

**Reply by kylehoang92** [reply]
Hi @magiziz, instead of setting `multiInjectedProviderDiscovery` to `false`, which disables EIP-6963 support, it may cause an issue where clicking "Connect Metamask" opens the OKX wallet prompt.

Can we include a filter for installed wallets based on the configured wallet list? 
Even if the wallet has been installed, it will not appear in the modal unless included in the `connectorsForWallets` configuration.

**Reply by DanielSinclair** [reply]
@kylehoang92 Is there a reason you wouldn't want to include an Installed wallet in your connector list? The idea is that RainbowKit will support all EIP-6963 wallets, so a user can connect no matter the wallet list configuration.

**Reply by kylehoang92** [reply]
Hi @DanielSinclair, my dapp requires the use of a custom chain. Some wallets, such as Phantom wallet, do not support adding custom networks. Additionally, others like Keplr wallet are not supported by rainbowkit. Users may install these extensions to access other dapps. However, when they visit my site and try to connect their wallets by clicking on them, they encounter connection issues.

**Reply by kylehoang92** [reply]
I want to display only the wallets I have added to `connectorsForWallets`. To achieve this, I can disable `multiInjectedProviderDiscovery`. However, there may be a connection issue as noted in the first comment. It would be ideal if there is an option to utilize EIP-6963 benefits while selectively showcasing supported wallets.

**Reply by DanielSinclair** [reply]
I see. That's valid. I think we would actually want to introduce this behavior into EIP-6963 as it currently lacks the ability for the wallet to advertise their supported chains. The way to approach this for now would be to allow a user to connect any wallet, attempt to switch to the appropriate chain with `useSwitchChain`. Wagmi will throw an error if the chain is unavailable and the wallet lacks support for Custom Chains. When this errors occurs, advise the user that their wallet isn't supported, and provide valid options. Alternatively, you could use our `WalletButton` API and build a custom connection flow that only supports specific wallets. As far as Keplr Wallet, we can definitely add support for that. If you're familair, do you mind opening a PR to add support? You can reference this PR: https://github.com/rainbow-me/rainbowkit/pull/2156/files

**Reply by kylehoang92** [reply]
Hi @DanielSinclair, thank you for the suggestion. Unfortunately, when I have multiple extensions installed on my browser and set multiInjectedProviderDiscovery to true, using WalletButton opens the wrong extension prompt.

**Reply by DanielSinclair** [reply]
@kylehoang92 Which wallet were you seeing that routing issues with? Some unfortunately still haven't fully adopted EIP-6963 which would prevent that. We have manual overrides as well for EIP-1193 and I can try to add one for that specific wallet

**Reply by kylehoang92** [reply]
@DanielSinclair I am having trouble with my MetaMask wallet, as it is being overridden by the OKX, Biget, or Trust wallets. I cannot connect to MetaMask properly.

**Reply by GraduErd** [reply]
@kylehoang92 did you find solution? with multiInjectedProviderDiscovery to false MetaMask wallet, as it is being overridden by the OKX , with true it shows unwanted wallets on installed parts i couldn't find solution

**Comment by Adebesin-Cell** [comment]
Hi @magiziz,

I've been trying to add a custom wallet - Magic. You can find the dedicated wallet connector definition [here](https://github.com/magiclabs/wagmi-magic-connector/blob/main/src/lib/connectors/dedicatedWalletConnector.ts). However, for some reason, it never shows up in the wallet list. It works fine in version 1.x but doesn't appear in version 2.x. Is there something additional that needs to be done for a custom wallet to show up?

Below are the relevant code snippet and screenshots:
<img width="748" alt="Screenshot 2024-08-20 at 5 09 55 PM" src="https://github.com/user-attachments/assets/bbbe4e3a-dd96-4834-9f39-f9caea77b748">

<img width="744" alt="Screenshot 2024-08-20 at 5 12 58 PM" src="https://github.com/user-attachments/assets/2932840c-1d7e-443a-b3ea-678e2d44116b">


```ts
export const rainbowMagicConnector = ({ chain }: { chain: Chain }): Wallet => ({
  id: 'magic',
  name: 'Magic',
  iconUrl: 'https://svgshare.com/i/pXA.svg',
  iconBackground: 'white',
  installed: true,
  rdns: 'io.magic',

  createConnector: () =>
    dedicatedWalletConnector({
      chains: [chain],
      options: {
        customLogo: 'https://iq.social/images/iq-logo.svg',
        apiKey: env.NEXT_PUBLIC_MAGIC_API_KEY,
        accentColor: '#ea3b87',
        oauthOptions: {
          providers: ['google', 'facebook', 'twitter', 'discord'],
        },
        isDarkMode: false,
        magicSdkConfiguration: {
          network: {
            chainId: chain.id,
            rpcUrl: chain.rpcUrls.default.http[0],
          },
        },
      },
    }),
})
```

Thanks in advance for any guidance!

**Reply by DanielSinclair** [reply]
@Adebesin-Cell Can you create a new GitHub issue for this?

**Reply by Adebesin-Cell** [reply]
For sure!
